src/Engine/Spells/CastSpellInfo.cpp

```diff
@@ -3313,40 +3295,63 @@ size_t PushCastSpellInfo(SPELL_TYPE uSpellID, uint16_t uPlayerID,
 }
 
 /**
+ * Remove all targeted spells from spell queue.
+ *
  * @offset 0x00427D48
  */
-void CastSpellInfoHelpers::Cancel_Spell_Cast_In_Progress() {  // reset failed/cancelled spell
+void CastSpellInfoHelpers::CancelSpellCastInProgress() {
+    bool targeted_spell_canceled = false;
+
     for (size_t i = 0; i < CastSpellInfoCount; i++) {
         if (pCastSpellInfo[i].uSpellID != SPELL_NONE &&
             pCastSpellInfo[i].uFlags & ON_CAST_CastingInProgress) {
+            // Only one targeted spell can exist in queue.
+            assert(!targeted_spell_canceled);
+
             pCastSpellInfo[i].uSpellID = SPELL_NONE;
 
             if (pGUIWindow_CastTargetedSpell) {
+                // TODO: where object is deleted?
                 pGUIWindow_CastTargetedSpell->Release();
                 pGUIWindow_CastTargetedSpell = nullptr;
             }
             mouse->SetCursorImage("MICON1");
             GameUI_StatusBar_Update(true);
             _50C9A0_IsEnchantingInProgress = 0;
             back_to_game();
+
+            targeted_spell_canceled = true;
         }
     }
 }
 
 /**
+ * General function that registers spell or skill implemented
+ * through spell casting mechanism to be cast/performed later.
+ *
+ * Actual casting will be performed with event queue processing.
+ *
+ * If spell is targeted then corresponding target mode is entered
+ * and actual casting will be performed after target is picked.
+ * Registered targeted spells will have one of the flags
+ * listed in ON_CAST_CastingInProgress and this flag will be removed
+ * in event queue if correct target is picked.
+ *
  * @offset 0x0042777D
  */
-void _42777D_CastSpell_UseWand_ShootArrow(SPELL_TYPE spell,
-                                          unsigned int uPlayerID,
-                                          PLAYER_SKILL skill_value, SpellCastFlags flags,
-                                          int a6) {
+void RegisterSpellOrSpellLikeSkill(SPELL_TYPE spell,
```

Comment:
I'd suggest naming this function `pushSpellOrRangedAttack`.

Using `register` in the name implies some kind of permanence of the action, while what's really happening under the hood in this function is just a call to `PushCastSpellInfo`.